contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -482,6 +490,41 @@ contract MixinMatchOrders is
             matchedFillResults.profitInRightMakerAsset
         );
 
+        // Pay protocol fees
+        if (staking != address(0)) {
+            // matchedFillResults.left.protocolFeePaid == matchedFillResults.right.protocolFeePaid
+            // so we only use matchedFillResults.left.protocolFeePaid as a gas optimization.
+            uint256 protocolFee = matchedFillResults.left.protocolFeePaid;
+
+            // Construct an array of makers and fee amounts so that the staking contract will only need to be called once.
+            address[] memory makers = new address[](2);
+            makers[0] = leftOrder.makerAddress;
+            makers[1] = rightOrder.makerAddress;
+            uint256[] memory fees = new uint256[](2);
+            fees[0] = protocolFee;
+            fees[1] = protocolFee;
+
+            // If sufficient ether was sent to the contract, the protocol fee should be paid in ETH.
+            // Otherwise the fee should be paid in WETH.
+            if (address(this).balance >= 2 * protocolFee) {
+                // Forward the protocol fees
+                IStaking(staking).batchPayProtocolFees.value(2 * protocolFee)(makers, fees);
+            } else {
+                // Transfer the weth from the takerAddress.
+                // Note: `_dispatchTransferFrom` is only called once as a gas optimization.
+                _dispatchTransferFrom(
+                    leftOrderHash,
+                    WETH_ASSET_DATA,
+                    takerAddress,
+                    staking,
+                    2 * protocolFee
+                );
+
+                // Attribute the protocol fees to the maker addresses
+                IStaking(staking).batchRecordProtocolFees(makers, fees);
```

Comment:
Same as above - let's put the WETH `transferFrom` call in `batchRecordProtocolFees`.

Reply:
I also wonder if we should just remove these batch methods for the sake of simplicity. We're only really saving ~700 gas here, and this won't be used anywhere else in the `Exchange` contract.